sdk/Sdk/FunctionMetadataGenerator.cs

```diff
@@ -42,44 +43,33 @@ public FunctionMetadataGenerator(Action<TraceLevel, string, string> log)
             }
         }
 
-        public IEnumerable<SdkFunctionMetadata> GenerateFunctionMetadata(string assemblyPath, IEnumerable<string> referencePaths)
+        public IEnumerable<SdkFunctionMetadata> GenerateFunctionMetadata(string assemblyPath, IEnumerable<ITaskItem> referencePaths)
         {
-            string sourcePath = Path.GetDirectoryName(assemblyPath);
+            var targetAssemblies = new List<string>() { assemblyPath };
 
-            var targetAssemblies = new List<string>(Directory.GetFiles(sourcePath, "*.dll"));
+            // We don't need to scan assemblies that come from a framework (.NET, AspNetCore), they won't have functions types in them.
+            targetAssemblies.AddRange(referencePaths.Where(x => x.GetMetadata("FrameworkReferenceName") == "").Select(x => x.ItemSpec));
+            _logger.LogMessage($"Found {targetAssemblies.Count} assemblies to evaluate.");
 
-            if (!assemblyPath.EndsWith(".dll"))
+            var resolver = new DefaultAssemblyResolver();
+            foreach (string referencePath in referencePaths.Select(p => Path.GetDirectoryName(p.ItemSpec)).Distinct())
```

Comment:
`referencePaths` is being enumerated more than once. Should we `ToArray` once and use that result in both places?

Reply:
Are you worried about the parameter `IEnumerable<ITaskItem> referencePaths` being enumerated twice? Or is one of the two projections I have being enumerated twice?

For the former, a caller may still supply a concrete list. In which case `ToArray` will be overhead we don't want. At build it will be a concrete list, so our only concern would be during a test. Which I don't want to change this API for.

For the later, the 2 projections of this parameter are unique and enumerated only once each. A concrete list will not benefit.